File: packages/shared-types/src/safety.ts
/**
 * Safety types — reports and blocks.
 *
 * Report: POST /reports
 * Block:  POST /blocks/{userId}
 *         DELETE /blocks/{userId}
 *         GET /blocks
 */

import type { ISODateString, UUID } from './common';

export type ReportReason =
  | 'spam'
  | 'inappropriate'
  | 'fake'
  | 'harassment'
  | 'other';

/**
 * Report target kind. Matches the backend `target_type` column:
 *   - "user"  → report is about a user (reported_id required)
 *   - "event" → report is about an event (target_event_id required)
 */
export type ReportTargetType = 'user' | 'event';

/**
 * Moderation lifecycle of a report. Matches the backend `status` column.
 * Only "actioned" feeds Honor; the other states are inert.
 */
export type ReportStatus =
  | 'submitted'
  | 'reviewed'
  | 'dismissed'
  | 'actioned';

export interface CreateReportRequest {
  reportedUserId: UUID;
  reason: ReportReason;
  context?: string;
}

export interface ReportResponse {
  id: UUID;
  reporterId: UUID;
  reportedId: UUID;
  reason: ReportReason;
  context?: string;
  createdAt: ISODateString;
}

/**
 * Response shape for `GET /reports/mine` — a paginated list of the
 * caller's submitted reports. `total` mirrors the block-list pattern.
 */
export interface ReportListResponse {
  items: ReportResponse[];
  total: number;
}

export interface BlockResponse {
  id: UUID;
  blockerId: UUID;
  blockedId: UUID;
  createdAt: ISODateString;
}

export interface BlockListResponse {
  items: BlockResponse[];
  total: number;
}

/**
 * Client-side helper shape for issuing a block. The block endpoint
 * uses `POST /blocks/{blockedUserId}` (path param only), so this is a
 * small request object the mobile client can pass around without
 * stringifying a bare UUID.
 */
export interface CreateBlockRequest {
  blockedUserId: UUID;
}

File: apps/api/app/schemas/safety.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

# Reason vocab keeps legacy V1 values and adds V1.1 entries so existing
# callers (Chat → user report) and new event-side flows both work.
ReportReason = Literal[
    # Legacy
    "spam",
    "inappropriate",
    "fake",
    "harassment",
    "other",
    # V1.1
    "no_show",
    "fraud_or_scam",
    "inappropriate_chat",
    "fake_profile",
    "unsafe_behavior",
]
ReportTargetType = Literal["user", "event"]
ReportStatus = Literal["submitted", "reviewed", "dismissed", "actioned"]


class CreateReportRequest(BaseModel):
    # Default to "user" so legacy callers omitting target_type still work.
    target_type: ReportTargetType = "user"
    # Legacy field name preserved for user reports.
    reported_user_id: UUID | None = None
    target_event_id: UUID | None = None
    reason: ReportReason
    context: str | None = Field(default=None, max_length=1000)
    # Status is server-controlled. The field is deliberately NOT
    # exposed on the request schema so clients cannot create
    # already-actioned reports.


class ReportResponse(BaseModel):
    id: UUID
    reporter_id: UUID
    target_type: str = "user"
    reported_id: UUID | None = None
    target_event_id: UUID | None = None
    reason: str
    context: str | None = None
    status: str = "submitted"
    created_at: datetime

    model_config = {"from_attributes": True}


class BlockResponse(BaseModel):
    id: UUID
    blocker_id: UUID
    blocked_id: UUID
    created_at: datetime

    model_config = {"from_attributes": True}


class BlockListResponse(BaseModel):
    items: list[BlockResponse]
    total: int


class ReportListResponse(BaseModel):
    """Response shape for GET /reports/mine — mirrors BlockListResponse."""

    items: list[ReportResponse]
    total: int
